### PR TITLE
remove replication clients handling out of vocbase code

### PR DESCRIPTION
**Pull request by jsteemann**

### CODE REVIEW

**Comment by jsteemann** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/3917/

**Comment by jsteemann** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/3918/

File: arangod/Aql/QueryRegistry.h [hunk 1]
```diff
@@ -109,6 +109,10 @@ class QueryRegistry {
    *        information.
    */
   void setNoLockHeaders(ExecutionEngine* engine) const;
+  
+ public:
+  /// @brief default timeout for holding read-locks on a collection
+  static constexpr double defaultReadLockTimeout = 900;
```
Comment (graetzer) [inline arangod/Aql/QueryRegistry.h]: shouldn't this be defined somewhere else ?

File: arangod/MMFiles/MMFilesEngine.cpp [hunk 2]
```diff
@@ -3424,23 +3424,7 @@ Result MMFilesEngine::createLoggerState(TRI_vocbase_t* vocbase, VPackBuilder& bu
   // "clients" part
   builder.add("clients", VPackValue(VPackValueType::Array));  // open
   if (vocbase != nullptr) {                                   // add clients
-    auto allClients = vocbase->getReplicationClients();
```
Comment (graetzer) [inline arangod/MMFiles/MMFilesEngine.cpp]: is this shown in the UI somewhere ?

File: arangod/Replication/ReplicationApplier.cpp [hunk 3]
```diff
@@ -26,6 +26,7 @@
 #include "Basics/Mutex.h"
 #include "Basics/MutexLocker.h"
 #include "Basics/ReadLocker.h"
+#include "Basics/StringUtils.h"
```
Comment (graetzer) [inline arangod/Replication/ReplicationApplier.cpp]: whats the include for

File: arangod/Replication/ReplicationClients.cpp [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+    } else {
+      ++it;
+    }
+  }
+}
+
+/// @brief return the lowest lastServedTick value for all clients
+/// returns UINT64_MAX in case no clients are registered
+uint64_t ReplicationClientsProgressTracker::lowestServedValue() const {
+  uint64_t value = UINT64_MAX;
```
Comment (graetzer) [inline arangod/Replication/ReplicationClients.cpp]: is uint64max a valid return value

File: arangod/RocksDBEngine/RocksDBRestReplicationHandler.cpp [hunk 5]
```diff
@@ -132,21 +126,16 @@ void RocksDBRestReplicationHandler::handleCommandBatch() {
...
-          << "no serverId parameter found in request to " << _request->fullUrl();
-    }
-
-    TRI_server_id_t serverId = id;  // just use context id as fallback
-    if (!value.empty() && value != "none") {
-      serverId = static_cast<TRI_server_id_t>(StringUtils::uint64(value));
+    std::string clientId = _request->value("serverId", found);
+    if (clientId.empty() || clientId == "none") {
+      // use context id as a fallback
+      clientId = std::to_string(id);
```
Comment (graetzer) [inline arangod/RocksDBEngine/RocksDBRestReplicationHandler.cpp]: I thini the you also implement this in the ReplicationContext above ?

File: arangod/Aql/QueryRegistry.h (hunk 1 above)
Comment (jsteemann) [reply arangod/Aql/QueryRegistry.h]: Ok, where does it belong then? Any suggestions?

File: arangod/MMFiles/MMFilesEngine.cpp (hunk 2 above)
Comment (jsteemann) [reply arangod/MMFiles/MMFilesEngine.cpp]: Yes, the web UI shows the connected clients in the "REPLICATION" tab on a master

File: arangod/Replication/ReplicationApplier.cpp (hunk 3 above)
Comment (jsteemann) [reply arangod/Replication/ReplicationApplier.cpp]: I had to add this so the branch actually compiled.
I removed an include of the StringUtils and a few other files from a central header file, which to led lots of compile errors. So I had to include the StringUtils in all the places where they are actually needed.
But it seems afterwards I removed the code from ReplicationApplier.cpp that was using the StringUtils functionality. So I will remove the include here now.

File: arangod/Replication/ReplicationClients.cpp (hunk 4 above)
Comment (jsteemann) [reply arangod/Replication/ReplicationClients.cpp]: Yes, the comments for this API explicitly state that.

File: arangod/Replication/ReplicationApplier.cpp (hunk 3 above)
Comment (jsteemann) [reply arangod/Replication/ReplicationApplier.cpp]: Correcting myself: the StringUtils functionality is still used in the ReplicationApplier, so the include must stay there in order to compile things.

File: arangod/Replication/ReplicationClients.cpp (hunk 4 above)
Comment (jsteemann) [reply arangod/Replication/ReplicationClients.cpp]: the call site will also ignore UINT64_MAX here, as the code there is (RocksDBBackgroundThread.cpp):
```
 92           // lowestServedValue will return the lowest of the lastServedTick values stored,
 93           // or UINT64_MAX if no clients are registered
 94           minTick = std::min(minTick, vocbase.replicationClients().lowestServedValue());
```

Comment (jsteemann) [reply arangod/Replication/ReplicationClients.cpp]: So returning UINT64_MAX shouldn't do any harm.

File: arangod/RocksDBEngine/RocksDBRestReplicationHandler.cpp (hunk 5 above)
Comment (jsteemann) [reply arangod/RocksDBEngine/RocksDBRestReplicationHandler.cpp]: Ok.

File: arangod/Aql/QueryRegistry.h (hunk 1 above)
Comment (jsteemann) [reply arangod/Aql/QueryRegistry.h]: I actually don't feel confident that this change was a good idea. I am reverting this part and setting the default timeout back to 0.0, which will use some default timeout (whatever this is).

**Comment by jsteemann** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/3941/

**Pull request merged**